src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java

```diff
@@ -1,437 +1,49 @@
-      } finally {
-        data.releaseReadLock();
-      }
-      currentThreads = getThreadsToUse((System.currentTimeMillis() - startTime), startMemory);
-      try {
-        // make sure all workers are using the same data
-        newData.acquireReadLock();
-        int i = 0;
-        // we are already in 1 executor thread, so we have MAX_THREADS-1 threads left to use
-        if (currentThreads <= 2 || MAX_THREADS <= 2) {
-          // if 2 or fewer threads, do not multi-thread the copying (we have already copied it once above, so at most
-          // only 1 more copy to
-          // make)
-          while (cancelCurrentOperation >= 0 && i < currentThreads) {
-            // the last one will use our already copied data from above, without copying it again
-            workers.add(new OddsCalculator(newData, (currentThreads == ++i)));
-          }
-        } else { // multi-thread our copying, cus why the heck not (it increases the speed of copying by about double)
-          final CountDownLatch workerLatch = new CountDownLatch(currentThreads - 1);
-          while (i < (currentThreads - 1)) {
-            ++i;
-            executor.submit(() -> {
-              if (cancelCurrentOperation >= 0) {
-                workers.add(new OddsCalculator(newData, false));
-              }
-              workerLatch.countDown();
-            });
-          }
-          // the last one will use our already copied data from above, without copying it again
-          workers.add(new OddsCalculator(newData, true));
-          try {
-            workerLatch.await();
-          } catch (final InterruptedException e) {
-            Thread.currentThread().interrupt();
-          }
-        }
-      } finally {
-        newData.releaseReadLock();
-      }
-    }
-    if (cancelCurrentOperation < 0 || data == null) {
-      // we could have cancelled while setting data, so clear the workers again if so
-      workers.clear();
-      isDataSet = false;
-    } else {
-      // should make sure that all workers have their game data set before we can call calculate and other things
-      isDataSet = true;
-      notifyListenersGameDataIsSet();
-    }
-    // allow setting new data to take place if it is waiting on us
-    latchWorkerThreadsCreation.countDown();
-    // allow calcing and other stuff to go ahead
-    latchSetData.countDown();
-    logger.fine("Initialized worker thread pool with size: " + workers.size());
-  }
-
-  @Override
-  public void shutdown() {
-    isShutDown = true;
-    cancelCurrentOperation = Integer.MIN_VALUE / 2;
-    cancel();
     executor.shutdown();
-    synchronized (listeners) {
-      listeners.clear();
-    }
-  }
-
-  private void awaitLatch() {
     try {
-      // there is a small chance calculate or setCalculateData or something could be called in between calls to
-      // setGameData
-      latchSetData.await();
+      executor.awaitTermination(15, TimeUnit.SECONDS);
```

Comment:
Is there any reasoning why this exact timeout has been chosen?
I'd suggest using using the boolean value returned by this method somewhere, ~to indicate the awaitTermination method has exceeded the timeout when it does~ to print a warning or something similar, that this method is taking way too long to execute

Reply:
a676536 adds some reasoning. Essentially 15s is beyond a very long time to wait for a calc, most are down in 2 to 5 seconds.

The error handling could be yet better, but at least we are timing out.

Reply:
I'd suggest using an ExecutorCompletionService here, which essentially wraps an executor service and allows you to retrieve future objects, and allows for easier error handling.
Essentially what I did in #2558

Reply:
I don't think we would want to even show an error dialog if we hit the timeout. The handling for now would be to cut the calc and display a result. If there were any results at all, there would be some stats. If we had error handling to show an error, ideally it would be done in background via network request to server for logging.

We could instead show a big error message to a player while playing the game. If they get zero results they'll likely file a bug report or move on. An error message even though there are results is more likely to cause the player confusion, eg: they may wonder if they should restart the game.

Given those scenarios, just cutting the calc at 15 seconds and not doing any further error handling is a reasonable scope and an improvement over the previous potential infinite wait. Hence I'm not inclined to update the error handling further for this scenario.